crates/matrix-sdk-indexeddb/src/crypto_store/mod.rs

```diff
@@ -1189,6 +1194,110 @@ impl Drop for IndexeddbCryptoStore {
     }
 }
 
+/// Fetch items from an object store in batches, transform each item using
+/// the supplied function, and stuff the transformed items into a single
+/// vector to return.
+async fn fetch_from_object_store_batched<R, F>(
+    object_store: IdbObjectStore<'_>,
+    f: F,
+    batch_size: usize,
+) -> Result<Vec<R>>
+where
+    F: Fn(JsValue) -> Result<R>,
+{
+    enum NextBatch {
+        First,          // In the first batch we don't start from a key
+        Later(JsValue), // In later batches we have a starting key
+    }
+
+    let mut result = Vec::new();
+    let mut next_batch = NextBatch::First;
+    let mut batch_n = 0;
+
+    loop {
+        debug!("Fetching Indexed DB records starting from {}", batch_n * batch_size);
+
+        // See https://github.com/Alorel/rust-indexed-db/issues/31 - we
+        // would like to use `get_all_with_key_and_limit` if it ever exists
+        // but for now we use a cursor and manually limit batch size.
+
+        // Get hold of a cursor for this batch
+        let cursor = match &next_batch {
+            NextBatch::First => object_store.open_cursor(),
+
+            NextBatch::Later(latest_key) => {
+                // We want a cursor starting after the last key we processed
+                let after_latest_key = IdbKeyRange::lower_bound_with_open(latest_key, true);
+
+                // This key should always be valid since we got hold of it
+                // within the same transaction.
+                let after_latest_key = after_latest_key.expect("Key was not valid!");
```

Comment:
Since this function already returns a `Result`, can we avoid the `expect()` and map to an error that we'd raise with `?` here?

Reply:
I tried to do that and couldn't find a good way to construct an `IndexeddbCryptoStoreError`, and also wasn't sure what error would be right. Would the correct thing be to change the error type, maybe adding an enum value to `IndexeddbCryptoStoreError`?

A separate but related question/wondering: ideally, we would be able to separate these batches into different transactions. That would mean that, in theory, a key might disappear during an export. What would be the correct behaviour if we "lost our place" during export because the key at the end of the last batch had disappeared? I will ask the crypto team whether we think this will ever happen.

Reply:
Yes, in general we shouldn't be afraid to add more specialized error types. I am for one guilty of taking shortcuts and reusing errors, but it's very fine to add new errors, and it should be painless in general thanks to the error derives we're using.

> That would mean that, in theory, a key might disappear during an export.

And the worst implication would be that we lose track of the cursor?

Reply:
> > That would mean that, in theory, a key might disappear during an export.
> 
> And the worst implication would be that we lose track of the cursor?

Yes, which I thought meant that we would fail mid-export but Rich pointed out to me that in indexeddb, the key range is literally a "key is > other key" thing, meaning if a key did disappear, we would still find the keys after it when we asked, so this would not even abort export.

Keys changing under us while we export is not particularly comfortable, but I am quite confident that any key that exists before the export starts and still exists after it ends will be included in the export, so I think that is a good enough invariant.

All of this discussion is actually about a future change, not this one anyway. This one preserves a single transaction for the whole export.

Reply:
OK, so I looked into this more, and I am fairly certain that this will only fail if we use the wrong _type_ as a key. We will never do this because we always pass in a key that was provided to us by `cursor.key()`. So I propose to update the comment here (7823dd8b4c) and leave it as an `expect()` since it should never happen.

Is "panicking when things that should never happen happen" in line with our normal practices @bnjbvr ?

Reply:
Usually the `panic`s I'm fine with are those that are caused by either one of two things:

- the platform isn't behaving as expected causing an error, and reporting that error would be impractical
- an invariant we just checked happens to not be hold

It seems that we're more or less in the first case here?

Reply:
Yes, the platform would have to either:

a) not accept a `&str`-derived `JsValue` as a valid key, or
b) not return a valid key from `cursor.key()`

for this to panic, and both of those would be unexpected. I think reporting that error would needlessly complicate our error enum, so I propose to leave it as is if you are happy with it.